js/private/js_package.bzl

```diff
@@ -0,0 +1,102 @@
+"js_package rule"
+
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory_lib")
+load("@aspect_bazel_lib//lib:copy_directory.bzl", "copy_directory_action")
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@rules_nodejs//nodejs:providers.bzl", "DeclarationInfo")
+
+JsPackageInfo = provider(
+    doc = "A provider that carries the output directory (a TreeArtifact) of a js_package which contains the packages sources along with the package name and version",
+    fields = {
+        "label": "the label of the target the created this provider",
+        "package": "name of this node package",
+        "version": "version of this node package",
+        "directory": "the output directory (a TreeArtifact) that contains the package sources",
+    },
+)
+
+_DOC = """A rule that packages sources into a TreeArtifact or forwards a tree artifact and provides a JsPackageInfo.
+
+This target can be used as the src attribute to link_js_package.
+
+A DeclarationInfo is also provided so that the target can be used as an input to rules that expect one such as ts_project."""
+
+_ATTRS = dicts.add({
+    "src": attr.label(
+        doc = "A source directory or output directory to use for this package. For specifying a list of files, use `srcs` instead.",
+        allow_single_file = True,
+    ),
+    "package": attr.string(
+        # TODO: validate that name matches in an action if src is set
+        doc = "Must match the `name` field in the `package.json` file for this package.",
```

Comment:
if there's a package.json in the folder already, can you have `package_from = ":package.json"` like we did for `ts_version`?

Reply:
not as trivial as that since the package.json we want is in the tree artifact and we're not in a repository rule context where we can open a package.json file and set the package & version attributes of a build action